internal/fs/caching_test.go

```diff
@@ -432,3 +436,313 @@ func (t *CachingWithImplicitDirsTest) SymlinksAreTypeCached() {
+		entries, err := fusetesting.ReadDirPicky(getMultiMountBucketDir(bucketName))
+		AssertEq(nil, err)
+
+		ExpectThat(entries, ElementsAre())
+	}
+}
+
+func (t *MultiBucketMountCachingTest) FileCreatedRemotely() {
+	const name = "foo"
+	const contents = "taco"
+	bucket1MntDir := getMultiMountBucketDir(bucket1Name)
+	bucket2MntDir := getMultiMountBucketDir(bucket2Name)
+
+	var fi os.FileInfo
+
+	// Create an object in GCS.
+	_, err := storageutil.CreateObject(
+		ctx,
+		uncachedBuckets[bucket1Name],
+		name,
+		[]byte(contents))
+
+	AssertEq(nil, err)
+
+	// It should immediately show up in a listing.
+	entries, err := fusetesting.ReadDirPicky(bucket1MntDir)
+	AssertEq(nil, err)
+	AssertEq(1, len(entries))
+
+	fi = entries[0]
+	ExpectEq(name, fi.Name())
+	ExpectEq(len(contents), fi.Size())
+
+	// we should not be able to stat it in the bucket2 mount directory
+	_, err = os.Stat(path.Join(bucket2MntDir, name))
+	AssertNe(nil, err)
+	AssertThat(err, Error(HasSubstr("no such file or directory")))
+
+	// And we should be able to stat it in bucket1 mount directory.
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+
+	ExpectEq(name, fi.Name())
+	ExpectEq(len(contents), fi.Size())
+
+	// And read it.
+	b, err := os.ReadFile(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectEq(contents, string(b))
+
+	// And overwrite it, and read it back again.
+	err = os.WriteFile(path.Join(bucket1MntDir, name), []byte("burrito"), 0500)
+	AssertEq(nil, err)
+
+	b, err = os.ReadFile(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectEq("burrito", string(b))
+
+	AssertEq(nil, os.Remove(path.Join(bucket1MntDir, name)))
+}
+
+func (t *MultiBucketMountCachingTest) FileChangedRemotely() {
+	const name = "foo"
+	const contents = "taco"
+	var fi os.FileInfo
+	var err error
+	bucket1MntDir := getMultiMountBucketDir(bucket1Name)
+
+	// Create an object in GCS.
+	_, err = storageutil.CreateObject(
+		ctx,
+		uncachedBuckets[bucket1Name],
+		name,
+		[]byte(contents))
+
+	AssertEq(nil, err)
+
+	// Create a file via the file system.
+	err = os.WriteFile(path.Join(bucket1MntDir, name), []byte("taco"), 0500)
+	AssertEq(nil, err)
+
+	// Overwrite the object in GCS.
+	_, err = storageutil.CreateObject(
+		ctx,
+		uncachedBuckets[bucket1Name],
+		name,
+		[]byte("burrito"))
+
+	AssertEq(nil, err)
+
+	// Because we are caching, the file should still appear to be the local
+	// version.
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectEq(len("taco"), fi.Size())
+
+	// After the TTL elapses, we should see the new version.
+	cacheClock.AdvanceTime(ttl + time.Millisecond)
+
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectEq(len("burrito"), fi.Size())
+
+	// Reading should work as expected.
+	b, err := os.ReadFile(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectEq("burrito", string(b))
+
+	AssertEq(nil, os.Remove(path.Join(bucket1MntDir, name)))
```

Comment:
can we have a teardown instead of deleting in the test itself ?

Reply:
A tear-down method would require us to maintain a list of files/directories/objects in struct/global-vars to be deleted, in this case. I think it is better to do deferred deletions within the test code itself, which is easier to both maintain and read. So, I've made changes taking that path. 

Let me know if you differ on that.

Reply:
Never mind, added teardown method in MultiBucketMountCachingTest .